packages/firestore/src/util/async_queue.ts

```diff
@@ -463,23 +463,21 @@ export class AsyncQueue {
   }
 
   /**
-   * For Tests: Runs some or all delayed operations early.
+   * Runs some or all delayed operations early.
    *
-   * @param lastTimerId Delayed operations up to and including this TimerId will
-   *  be drained. Throws if no such operation exists. Pass TimerId.All to run
-   *  all delayed operations.
+   * @param timerId Delayed operations to run. Pass TimerId.All to run all
+   * delayed operations.
    * @returns a Promise that resolves once all operations have been run.
    */
-  runDelayedOperationsEarly(lastTimerId: TimerId): Promise<void> {
+  runDelayedOperationsEarly(timerId: TimerId): Promise<void> {
```

Comment:
This breaks the intent of what this test-only method was for. The idea was that you could fast-forward time until the delayed operation ran, including anything scheduled prior to it. In this interface, now you have to know exactly what you want to run, which makes this a much less effective tool for testing. In particular, this makes it really hard to use this for spec tests because we're no longer simulating the actual passage of time, but merely jumping ahead to this specific event. This makes it far less likely that we'll catch scheduling-related problems with tests.

If we really need to allow a scheduled operation to run early, can't we add this as a method to the `DelayedOperation`?

Reply:
I updated the code to return DelayedOperation instead of CancellablePromise, and added a skipDelay to ExponentialBackoff.

I feel that we might even get rid of the TimerIds with this change, since we now have direct access to the underlying operations - but that of course would remove the ability to run delayed operations in order. It might drastically simplify this code though.